File: packages/core/src/scripting/lib/list.ts
import {
  evaluate,
  executeLambda,
  ScriptError,
  OpcodeDefinition,
} from "../interpreter";

export const ListLibrary: Record<string, OpcodeDefinition> = {
  "list.new": {
    metadata: {
      label: "List",
      category: "list",
      description: "Create a list",
      // TODO: List of slots
      slots: [],
    },
    handler: async (args, ctx) => {
      const result = [];
      for (const arg of args) {
        result.push(await evaluate(arg, ctx));
      }
      return result;
    },
  },
  "list.len": {
    metadata: {
      label: "List Length",
      category: "list",
      description: "Get list length",
      slots: [{ name: "List", type: "block" }],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      if (!Array.isArray(list)) return 0;
      return list.length;
    },
  },

  "list.empty": {
    metadata: {
      label: "Is Empty",
      category: "list",
      description: "Check if list is empty",
      slots: [{ name: "List", type: "block" }],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      if (!Array.isArray(list)) return true;
      return list.length === 0;
    },
  },

  "list.get": {
    metadata: {
      label: "Get Item",
      category: "list",
      description: "Get item at index",
      slots: [
        { name: "List", type: "block" },
        { name: "Index", type: "number" },
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const index = await evaluate(args[1], ctx);
      if (!Array.isArray(list)) return null;
      return list[index];
    },
  },

  "list.set": {
    metadata: {
      label: "Set Item",
      category: "list",
      description: "Set item at index",
      slots: [
        { name: "List", type: "block" },
        { name: "Index", type: "number" },
        { name: "Value", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const index = await evaluate(args[1], ctx);
      const val = await evaluate(args[2], ctx);
      if (!Array.isArray(list)) return null;
      list[index] = val;
      return val;
    },
  },

  "list.push": {
    metadata: {
      label: "Push",
      category: "list",
      description: "Add item to end",
      slots: [
        { name: "List", type: "block" },
        { name: "Value", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const val = await evaluate(args[1], ctx);
      if (!Array.isArray(list)) return null;
      list.push(val);
      return list.length;
    },
  },

  "list.pop": {
    metadata: {
      label: "Pop",
      category: "list",
      description: "Remove item from end",
      slots: [{ name: "List", type: "block" }],
    },
    handler: async (args, ctx) => {
      if (args.length !== 1) {
        throw new ScriptError("list.pop: expected 1 argument");
      }
      const [listExpr] = args;
      const list = await evaluate(listExpr, ctx);
      if (!Array.isArray(list)) return null;
      return list.pop();
    },
  },

  "list.unshift": {
    metadata: {
      label: "Unshift",
      category: "list",
      description: "Add item to start",
      slots: [
        { name: "List", type: "block" },
        { name: "Value", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const val = await evaluate(args[1], ctx);
      if (!Array.isArray(list)) return null;
      list.unshift(val);
      return list.length;
    },
  },

  "list.shift": {
    metadata: {
      label: "Shift",
      category: "list",
      description: "Remove item from start",
      slots: [{ name: "List", type: "block" }],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      if (!Array.isArray(list)) return null;
      return list.shift();
    },
  },

  "list.slice": {
    metadata: {
      label: "Slice List",
      category: "list",
      description: "Extract part of list",
      slots: [
        { name: "List", type: "block" },
        { name: "Start", type: "number" },
        { name: "End", type: "number", default: null },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length < 2 || args.length > 3) {
        throw new ScriptError("list.slice: expected 2 or 3 arguments");
      }
      const [listExpr, startExpr, endExpr] = args;
      const list = await evaluate(listExpr, ctx);
      const start = await evaluate(startExpr, ctx);
      const end = endExpr ? await evaluate(endExpr, ctx) : undefined;
      if (!Array.isArray(list) || typeof start !== "number") return [];
      return list.slice(start, end);
    },
  },

  "list.splice": {
    metadata: {
      label: "Splice List",
      category: "list",
      description: "Remove/Replace items",
      slots: [
        { name: "List", type: "block" },
        { name: "Start", type: "number" },
        { name: "Delete Count", type: "number" },
        { name: "Items", type: "block" }, // Variadic
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const start = await evaluate(args[1], ctx);
      const deleteCount = await evaluate(args[2], ctx);
      // Remaining args are items to insert
      const items = [];
      for (let i = 3; i < args.length; i++) {
        items.push(await evaluate(args[i], ctx));
      }
      if (!Array.isArray(list)) return [];
      return list.splice(start, deleteCount, ...items);
    },
  },

  "list.concat": {
    metadata: {
      label: "Concat Lists",
      category: "list",
      description: "Concatenate lists",
      slots: [
        { name: "List 1", type: "block" },
        { name: "List 2", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length !== 2) {
        throw new ScriptError("list.concat: expected 2 arguments");
      }
      const [list1Expr, list2Expr] = args;
      const list1 = await evaluate(list1Expr, ctx);
      const list2 = await evaluate(list2Expr, ctx);
      if (!Array.isArray(list1) || !Array.isArray(list2)) return [];
      return list1.concat(list2);
    },
  },

  "list.includes": {
    metadata: {
      label: "List Includes",
      category: "list",
      description: "Check if list includes item",
      slots: [
        { name: "List", type: "block" },
        { name: "Value", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const val = await evaluate(args[1], ctx);
      if (!Array.isArray(list)) return false;
      return list.includes(val);
    },
  },

  "list.reverse": {
    metadata: {
      label: "Reverse List",
      category: "list",
      description: "Reverse list order",
      slots: [{ name: "List", type: "block" }],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      if (!Array.isArray(list)) return list;
      return list.reverse();
    },
  },

  "list.sort": {
    metadata: {
      label: "Sort List",
      category: "list",
      description: "Sort list",
      slots: [{ name: "List", type: "block" }],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      if (!Array.isArray(list)) return list;
      return list.sort();
    },
  },

  "list.join": {
    metadata: {
      label: "Join List",
      category: "list",
      description: "Join list to string",
      slots: [
        { name: "List", type: "block" },
        { name: "Separator", type: "string" },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length !== 2) {
        throw new ScriptError("list.join: expected 2 arguments");
      }
      const [listExpr, sepExpr] = args;
      const list = await evaluate(listExpr, ctx);
      const sep = await evaluate(sepExpr, ctx);
      if (!Array.isArray(list) || typeof sep !== "string") return "";
      return list.join(sep);
    },
  },

  "list.find": {
    metadata: {
      label: "Find Item",
      category: "list",
      description: "Find item in list",
      slots: [
        { name: "List", type: "block" },
        { name: "Lambda", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length !== 2) {
        throw new ScriptError("list.find: expected 2 arguments");
      }
      const [listExpr, funcExpr] = args;
      const list = await evaluate(listExpr, ctx);
      const func = await evaluate(funcExpr, ctx);

      if (!Array.isArray(list) || !func || func.type !== "lambda") return null;

      for (const item of list) {
        if (await executeLambda(func, [item], ctx)) {
          return item;
        }
      }
      return null;
    },
  },

  "list.map": {
    metadata: {
      label: "Map List",
      category: "list",
      description: "Map list items",
      slots: [
        { name: "List", type: "block" },
        { name: "Lambda", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length !== 2) {
        throw new ScriptError("list.map: expected 2 arguments");
      }
      const [listExpr, funcExpr] = args;
      const list = await evaluate(listExpr, ctx);
      const func = await evaluate(funcExpr, ctx);

      if (!Array.isArray(list) || !func || func.type !== "lambda") return [];

      const result: unknown[] = [];
      for (const item of list) {
        result.push(await executeLambda(func, [item], ctx));
      }
      return result;
    },
  },

  "list.filter": {
    metadata: {
      label: "Filter List",
      category: "list",
      description: "Filter list items",
      slots: [
        { name: "List", type: "block" },
        { name: "Lambda", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length !== 2) {
        throw new ScriptError("list.filter: expected 2 arguments");
      }
      const [listExpr, funcExpr] = args;
      const list = await evaluate(listExpr, ctx);
      const func = await evaluate(funcExpr, ctx);

      if (!Array.isArray(list) || !func || func.type !== "lambda") return [];

      const result: unknown[] = [];
      for (const item of list) {
        if (await executeLambda(func, [item], ctx)) {
          result.push(item);
        }
      }
      return result;
    },
  },

  "list.reduce": {
    metadata: {
      label: "Reduce List",
      category: "list",
      description: "Reduce list items",
      slots: [
        { name: "List", type: "block" },
        { name: "Lambda", type: "block" },
        { name: "Init", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      if (args.length !== 3) {
        throw new ScriptError("list.reduce: expected 3 arguments");
      }
      const [listExpr, funcExpr, initExpr] = args;
      const list = await evaluate(listExpr, ctx);
      const func = await evaluate(funcExpr, ctx);
      let acc = await evaluate(initExpr, ctx);

      if (!Array.isArray(list) || !func || func.type !== "lambda") return acc;

      for (const item of list) {
        acc = await executeLambda(func, [acc, item], ctx);
      }
      return acc;
    },
  },

  "list.flatMap": {
    metadata: {
      label: "FlatMap List",
      category: "list",
      description: "FlatMap list items",
      slots: [
        { name: "List", type: "block" },
        { name: "Lambda", type: "block" },
      ],
    },
    handler: async (args, ctx) => {
      const list = await evaluate(args[0], ctx);
      const func = await evaluate(args[1], ctx);
      if (!Array.isArray(list)) return [];

      const result: unknown[] = [];
      for (const item of list) {
        const mapped = await executeLambda(func, [item], ctx);
        if (Array.isArray(mapped)) {
          result.push(...mapped);
        } else {
          result.push(mapped);
        }
      }
      return result;
    },
  },
};

File: crates/runtime/luajit/src/codegen/list.rs
//! list.* opcode compilation.

use super::{CompileError, compile_value};
use viwo_ir::SExpr;

/// Compile list.* opcodes. Returns None if opcode doesn't match.
pub fn compile_list(
    op: &str,
    args: &[SExpr],
    prefix: &str,
) -> Result<Option<String>, CompileError> {
    let result = match op {
        "list.new" => {
            if args.is_empty() {
                // Empty array needs metatable to serialize as [] not {}
                format!("{}setmetatable({{}}, __array_mt)", prefix)
            } else {
                let elements: Result<Vec<_>, _> =
                    args.iter().map(|a| compile_value(a, false)).collect();
                format!(
                    "{}setmetatable({{ {} }}, __array_mt)",
                    prefix,
                    elements?.join(", ")
                )
            }
        }

        "list.len" => {
            if args.is_empty() {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 1,
                    got: 0,
                });
            }
            let arg = compile_value(&args[0], false)?;
            format!("{}#{}", prefix, arg)
        }

        "list.get" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let index = compile_value(&args[1], false)?;
            // Support optional third argument as default value
            if args.len() >= 3 {
                let default = compile_value(&args[2], false)?;
                // Lua is 1-indexed, so add 1 to the index
                // Use `~= nil and ... or default` pattern for safe default
                format!(
                    "{}({}[{} + 1] ~= nil and {}[{} + 1] or {})",
                    prefix, list, index, list, index, default
                )
            } else {
                // Lua is 1-indexed, so add 1 to the index
                format!("{}{}[{} + 1]", prefix, list, index)
            }
        }

        "list.set" => {
            if args.len() < 3 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 3,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let index = compile_value(&args[1], false)?;
            let value = compile_value(&args[2], false)?;
            // Lua is 1-indexed
            format!("{}[{} + 1] = {}", list, index, value)
        }

        "list.push" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let value = compile_value(&args[1], false)?;
            format!("{}table.insert({}, {})", prefix, list, value)
        }

        "list.pop" => {
            if args.is_empty() {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 1,
                    got: 0,
                });
            }
            let list = compile_value(&args[0], false)?;
            format!("{}table.remove({})", prefix, list)
        }

        "list.map" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let func = compile_value(&args[1], false)?;
            format!(
                "{}(function(arr, fn) local r = {{}}; for i, v in ipairs(arr) do r[i] = fn(v, i - 1) end; return setmetatable(r, __array_mt) end)({}, {})",
                prefix, list, func
            )
        }

        "list.filter" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let func = compile_value(&args[1], false)?;
            format!(
                "{}(function(arr, fn) local r = {{}}; for i, v in ipairs(arr) do if fn(v, i - 1) then r[#r+1] = v end end; return setmetatable(r, __array_mt) end)({}, {})",
                prefix, list, func
            )
        }

        "list.reduce" => {
            if args.len() < 3 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 3,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let func = compile_value(&args[1], false)?;
            let init = compile_value(&args[2], false)?;
            format!(
                "{}(function(arr, fn, acc) for i, v in ipairs(arr) do acc = fn(acc, v, i - 1) end; return acc end)({}, {}, {})",
                prefix, list, func, init
            )
        }

        "list.find" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let func = compile_value(&args[1], false)?;
            format!(
                "{}(function(arr, fn) for i, v in ipairs(arr) do if fn(v, i - 1) then return v end end; return nil end)({}, {})",
                prefix, list, func
            )
        }

        "list.concat" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list1 = compile_value(&args[0], false)?;
            let list2 = compile_value(&args[1], false)?;
            format!(
                "{}(function(a, b) local r = {{}}; for _, v in ipairs(a) do r[#r+1] = v end; for _, v in ipairs(b) do r[#r+1] = v end; return setmetatable(r, __array_mt) end)({}, {})",
                prefix, list1, list2
            )
        }

        "list.slice" => {
            if args.len() < 3 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 3,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let start = compile_value(&args[1], false)?;
            let end = compile_value(&args[2], false)?;
            format!(
                "{}(function(arr, s, e) local r = {{}}; for i = s + 1, e do r[#r+1] = arr[i] end; return setmetatable(r, __array_mt) end)({}, {}, {})",
                prefix, list, start, end
            )
        }

        "list.empty" => {
            if args.is_empty() {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 1,
                    got: 0,
                });
            }
            let list = compile_value(&args[0], false)?;
            format!("{}(#{} == 0)", prefix, list)
        }

        "list.unshift" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let value = compile_value(&args[1], false)?;
            // table.insert with position 1 adds to front
            format!("{}table.insert({}, 1, {})", prefix, list, value)
        }

        "list.shift" => {
            if args.is_empty() {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 1,
                    got: 0,
                });
            }
            let list = compile_value(&args[0], false)?;
            // table.remove with position 1 removes from front
            format!("{}table.remove({}, 1)", prefix, list)
        }

        "list.includes" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let value = compile_value(&args[1], false)?;
            format!(
                "{}(function(arr, val) for _, v in ipairs(arr) do if v == val then return true end end; return false end)({}, {})",
                prefix, list, value
            )
        }

        "list.indexOf" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let value = compile_value(&args[1], false)?;
            format!(
                "{}(function(arr, val) for i, v in ipairs(arr) do if v == val then return i - 1 end end; return -1 end)({}, {})",
                prefix, list, value
            )
        }

        "list.reverse" => {
            if args.is_empty() {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 1,
                    got: 0,
                });
            }
            let list = compile_value(&args[0], false)?;
            // Mutates in place and returns the array
            format!(
                "{}(function(arr) local n = #arr; for i = 1, math.floor(n/2) do arr[i], arr[n-i+1] = arr[n-i+1], arr[i] end; return arr end)({})",
                prefix, list
            )
        }

        "list.sort" => {
            if args.is_empty() {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 1,
                    got: 0,
                });
            }
            let list = compile_value(&args[0], false)?;
            // table.sort mutates in place, return the array
            format!(
                "{}(function(arr) table.sort(arr); return arr end)({})",
                prefix, list
            )
        }

        "list.join" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let sep = compile_value(&args[1], false)?;
            format!("{}table.concat({}, {})", prefix, list, sep)
        }

        "list.splice" => {
            // list.splice(arr, start, deleteCount, ...items)
            if args.len() < 3 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 3,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let start = compile_value(&args[1], false)?;
            let delete_count = compile_value(&args[2], false)?;
            let items: Result<Vec<_>, _> =
                args[3..].iter().map(|a| compile_value(a, false)).collect();
            let items_str = items?.join(", ");
            // Returns removed elements, modifies array in place
            format!(
                "{}(function(arr, s, d, ...) local r = {{}}; local items = {{...}}; s = s + 1; for i = 1, d do if arr[s] then r[#r+1] = table.remove(arr, s) end end; for i = #items, 1, -1 do table.insert(arr, s, items[i]) end; return setmetatable(r, __array_mt) end)({}, {}, {}{})",
                prefix,
                list,
                start,
                delete_count,
                if items_str.is_empty() {
                    String::new()
                } else {
                    format!(", {}", items_str)
                }
            )
        }

        "list.flatMap" => {
            if args.len() < 2 {
                return Err(CompileError::InvalidArgCount {
                    opcode: op.to_string(),
                    expected: 2,
                    got: args.len(),
                });
            }
            let list = compile_value(&args[0], false)?;
            let func = compile_value(&args[1], false)?;
            format!(
                "{}(function(arr, fn) local r = {{}}; for i, v in ipairs(arr) do local res = fn(v, i - 1); if type(res) == 'table' then for _, item in ipairs(res) do r[#r+1] = item end else r[#r+1] = res end end; return setmetatable(r, __array_mt) end)({}, {})",
                prefix, list, func
            )
        }

        _ => return Ok(None),
    };

    Ok(Some(result))
}

#[cfg(test)]
mod tests {
    use super::super::compile;
    use viwo_ir::SExpr;

    #[test]
    fn test_new() {
        let expr = SExpr::call(
            "list.new",
            vec![
                SExpr::number(1).erase_type(),
                SExpr::number(2).erase_type(),
                SExpr::number(3).erase_type(),
            ],
        );
        assert_eq!(
            compile(&expr).unwrap(),
            "return setmetatable({ 1, 2, 3 }, __array_mt)"
        );
    }

    #[test]
    fn test_new_empty() {
        let expr = SExpr::call("list.new", vec![]);
        assert_eq!(
            compile(&expr).unwrap(),
            "return setmetatable({}, __array_mt)"
        );
    }

    #[test]
    fn test_get() {
        let expr = SExpr::call(
            "list.get",
            vec![
                SExpr::call(
                    "list.new",
                    vec![
                        SExpr::number(10).erase_type(),
                        SExpr::number(20).erase_type(),
                    ],
                ),
                SExpr::number(0).erase_type(),
            ],
        );
        // Lua is 1-indexed, so we add 1
        assert!(compile(&expr).unwrap().contains("[0 + 1]"));
    }

    #[test]
    fn test_get_with_default() {
        let expr = SExpr::call(
            "list.get",
            vec![
                SExpr::call("std.var", vec![SExpr::string("arr").erase_type()]),
                SExpr::number(5).erase_type(),
                SExpr::string("missing").erase_type(),
            ],
        );
        // Should use nil check with default value
        let code = compile(&expr).unwrap();
        assert!(code.contains("[5 + 1] ~= nil"));
        assert!(code.contains("\"missing\""));
    }

    #[test]
    fn test_len() {
        let expr = SExpr::call(
            "list.len",
            vec![SExpr::call(
                "list.new",
                vec![SExpr::number(1).erase_type(), SExpr::number(2).erase_type()],
            )],
        );
        assert!(compile(&expr).unwrap().starts_with("return #"));
    }

    #[test]
    fn test_push() {
        let expr = SExpr::call(
            "list.push",
            vec![
                SExpr::call("std.var", vec![SExpr::string("arr").erase_type()]),
                SExpr::number(5).erase_type(),
            ],
        );
        assert!(compile(&expr).unwrap().contains("table.insert"));
    }
}
